cli/command/telemetry_utils.go

```diff
@@ -9,11 +9,19 @@ import (
 
 	"github.com/docker/cli/cli/version"
 	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
+	"go.opentelemetry.io/otel"
 	"go.opentelemetry.io/otel/attribute"
 	"go.opentelemetry.io/otel/metric"
 )
 
+var DefaultOTELErrorHandler otel.ErrorHandler = otel.ErrorHandlerFunc(handleOTELError)
+
+func init() {
+	otel.SetErrorHandler(DefaultOTELErrorHandler)
+}
```

Comment:
I'm not a big fan of `init` funcs. Could we do this inside the main func instead?

Is `DefaultOTELErrHandler` necessary? Can't we do this instead
```suggestion
func init() {
	otel.SetErrorHandler(otel.ErrorHandlerFunc(handleOTELError))
}
```

Reply:
Yeah, also not a huge fan of `init`. If we _must_ keep this, we could even consider inlinining `handleOTELError`;

```suggestion
func init() {
	otel.SetErrorHandler(otel.ErrorHandlerFunc(func(err error) {
		logrus.WithError(err).Debug("otel error")
	}))
}
```

But I'd also be more in favour of having this up to the consumer to decide (and set the error-handler in their `main` or elsewhere).

Reply:
+1 to not having this as an `init`, unless we have strong arguments for it.

(but if we must do it that way, this is one of the less awful uses of `init` and I'd be okay with it 😅)

Reply:
The main reason is because this handler would also be standard for any plugins and I don't want to necessarily be too invasive in requiring those plugins to know to set the error handler.

I'm not a big fan of the init. I'll see if I can find a more common place to put this code.

The primary reason for `DefaultOTELErrorHandler` is because I also plan that this error handler is used by plugins too and I didn't want to leak the implementation. I can create the variable though and inline the function though since that won't really leak anything.

Reply:
Yeah, so I was looking at this from some perspectives;

- it's setting some defaults in an `init`, which may beccome part of imported code in other code-bases
- the default is to set up a debug-logger; performing logging as part of library code is often considered a bit of a "no-no"
- I'm also looking a the `logrus` import in this case, as I know we're trying to move away from hard-coding logrus as implementation for logging; in the moby code-base we started to use the `containerd/log` module (https://github.com/containerd/log), which is currently a wrapper around logrus, but could allow using that abstraction to move to a different implementation, as well as allows for implementations to attach a logger through the context (unfortunately a context-logger doesn't look to be a real option here due to the signature).

Reply:
This logging is meant to tie in with the `github.com/dockerr/cli/cli/debug` package which explicitly uses logrus for the debugging. What if I were to move the error handler to there, remove the `init()` function, and invoke `otel.SetErrorHandler` on both the `runDocker` function and inside of one of the plugin packages?

Reply:
Another option I was thinking about here could be something like this:

- create an ad-hoc func (eg `setDefaultOtelErrorHandler`) for setting this error handler instead of init, and make sure it only executes once;
- call this handler in both our "provider constructor" funcs, so:
  - in `func (cli *DockerCli) MeterProvider(..)`;
  - and in `func (cli *DockerCli) TracerProvider(..)`

Tiny bit of repetition, but it'd still be transparent to plugins as long as they actually use one of the given providers, and we can avoid the `init()` func which would be run even when no other otel bits are being used.

WDYT?

Reply:
I'm not the biggest fan of registering these with the provider constructors.

I did a bit of changes and moved the error handler to the debug package that specifically uses `logrus` so it's easier to find and change whenever that part gets refactored. I also changed the `SetErrorHandler` call to be invoked from the `plugin.Run` function and inside of `cmd/docker/main.go` directly so it's no longer in an `init()` function. I think this might solve the problems we were discussing although I don't know if these are the best places for these calls.